File: kcp.py
#!/usr/bin/env python3
# License: Public Domain
# Release: 0.8

import argparse, os, sys, subprocess, json
from urllib import request

search_head  = 'application/vnd.github.v3.text-match+json'
search_h_tp  = 'Accept'
search_base  = 'https://api.github.com/search/repositories?q={}+user:KaOS-Community-Packages'
url_base     = 'https://github.com/KaOS-Community-Packages/{}.git'
url_pkgbuild = 'https://raw.githubusercontent.com/KaOS-Community-Packages/{}/master/PKGBUILD'

def print_error(msg):
	print('\033[1;31m{}\033[m'.format(msg))

def question(msg, default_value=True):
	if default_value:
		choice_rep = '[Y/n]'
	else:
		choice_rep = '[y/N]'
	response = input('\033[1;33m{} {} \033[m'.format(msg, choice_rep)).lower()
	if not response:
		return default_value
	if response[0] == 'y':
		return True
	if response[0] == 'n':
		return False
	return default_value

def edit(file_name):
	os.system('$EDITOR {}'.format(file_name))

def check_user():
	if not os.geteuid():
		print_error("Don't launch this program as root!")
		sys.exit(1)

def edit_pkgbuild():
	if question('Do you want to edit PKGBUILD?'):
		edit('PKGBUILD')

def get_package(app):
	url = url_base.format(app)
	exe = subprocess.Popen(['git', 'clone', url])
	err = exe.wait()
	if err:
		sys.exit(err)

def launch_request(search, header = None):
	req = request.Request(search)
	if header:
		req.add_header(search_h_tp, header)
	try:
		result = request.urlopen(req)
		return result.read().decode()
	except:
		return ''

def get_version(result):
	pkgver, pkgrel = '', ''
	for l in result.split('\n'):
		e = l.strip()
		if e[:7] == 'pkgver=':
			pkgver = e[7:]
		elif e[:7] == 'pkgrel=':
			pkgrel = e[7:]
	if pkgver and pkgrel:
		return '{}-{}'.format(pkgver, pkgrel)
	return '<unknown>'

def check_installed(app):
	exe = subprocess.Popen(['pacman', '-Q', app], stdout=subprocess.PIPE, stderr=subprocess.PIPE)	
	if exe.wait():
		return ''
	return exe.stdout.read().decode().split()[1]

def search_package(app, fast):
	search = search_base.format(app)
	result = json.loads(launch_request(search, search_head))
	for a in result['items']:
		n, d, s = a['name'], a['description'], a['stargazers_count']
		i = check_installed(a['name'])
		if fast:
			if i:
				print('\033[1m{}\033[m \033[1;36m[installed: {}]\033[m \033[1;34m({})\033[m'.format(n, i, s))
			else:
				print('\033[1m{}\033[m\033 \033[1;34m({})\033[m'.format(n, s))
		else:
			v = get_version(launch_request(url_pkgbuild.format(n)))
			if i:
				if v == i:
					i = ' [installed]'
				else:
					i = ' [installed: {}]'.format(i)
			print('\033[1m{}\033[m \033[1;32m{}\033[m\033[1;36m{}\033[m \033[1;34m({})\033[m'.format(n, v, i, s))
		print('\t{}'.format(d))

def install_package(app, asdeps):
	os.chdir('/tmp')
	get_package(app)
	os.chdir('/tmp/{}'.format(app))
	edit_pkgbuild()
	cmd = 'makepkg -si'
	if asdeps:
		cmd += ' --asdeps'
	os.system(cmd)
	os.system('rm -rf {}'.format(os.getcwd()))

def build_args():
	parser = argparse.ArgumentParser(description='Tool in command-line for KaOS Community Packages')
	parser.add_argument('-v', '--version', help='print version', action='version', version='0.8')
	group = parser.add_mutually_exclusive_group()
	group.add_argument('-g', '--get', help='get needed files to build app', metavar='APP')
	group.add_argument('-s', '--search', help='search an app in KCP', metavar='APP')
	group.add_argument('-i', '--install', help='install an app in KCP', metavar='APP')
	parser.add_argument('--asdeps', help='install as a dependence', action='store_true', default=False)
	parser.add_argument('--fast', help='search without version', action='store_true', default=False)
	return parser

check_user()

parser = build_args()
if len(sys.argv) == 1:
	parser.print_help()
	sys.exit(0)

args = parser.parse_args()
if args.get:
	get_package(args.get)
elif args.search:
	search_package(args.search, args.fast)
elif args.install:
	install_package(args.install, args.asdeps)
else:
	parser.print_help()

File: kcp.go
package main

/*
 * Depends on goopt: https://github.com/droundy/goopt
 */

import (
	"encoding/json"
	"fmt"
	"goopt"
	"io/ioutil"
	"net/http"
	"os"
	"os/exec"
	"strings"
)

const (
	versionNumber  = "0.10"
	defaultEditor  = "vim"
	searchHead     = "application/vnd.github.v3.text-match+json"
	searchHeadType = "Accept"
	searchBase     = "https://api.github.com/search/repositories?q=%v+user:KaOS-Community-Packages+fork:true"
	urlBase        = "https://github.com/KaOS-Community-Packages/%v.git"
	urlPkgbuild    = "https://raw.githubusercontent.com/KaOS-Community-Packages/%v/master/PKGBUILD"
)

var editor string
var tmpDir string
var fGet, fInstall, fSearch *string
var fHelp, fVersion, fFast, fDeps *bool

func init() {
	editor = os.Getenv("EDITOR")
	if editor == "" {
		editor = defaultEditor
	}
	tmpDir = os.TempDir()

	fHelp = goopt.Flag([]string{"-h", "--help"}, []string{}, "show this help message and exit", "")
	fVersion = goopt.Flag([]string{"-v", "--version"}, []string{}, "print version", "")
	fGet = goopt.String([]string{"-g", "--get"}, "", "get needed files to build app")
	fSearch = goopt.String([]string{"-s", "--search"}, "", "search an app in KCP")
	fFast = goopt.Flag([]string{"--fast"}, []string{}, "search without version", "")
	fInstall = goopt.String([]string{"-i", "--install"}, "", "install an app in KCP")
	fDeps = goopt.Flag([]string{"--asdeps"}, []string{}, "install as a dependence", "")

	goopt.Version = "0.10"
	goopt.Summary = "Tool in command-line for KaOS Community Packages"

}

func printError(msg string) {
	fmt.Printf("\033[1;31m%v\033[m\n", msg)
}

func question(msg string, value bool) bool {
	var defaultVal string
	if value {
		defaultVal = "[Y/n]"
	} else {
		defaultVal = "[y/N]"
	}
	fmt.Printf("\033[1;33m%v %v \033[m", msg, defaultVal)
	var response string
	if _, err := fmt.Scanf("%v", &response); err != nil {
		return value
	}
	if len(response) == 0 {
		return value
	}
	response = strings.ToLower(response)
	if strings.HasPrefix(response, "y") {
		return true
	} else if strings.HasPrefix(response, "n") {
		return false
	}
	return value
}

func edit(filename string) {
	cmd := exec.Command(editor, filename)
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Run()
}

func checkUser() {
	if e := os.Geteuid(); e == 0 {
		printError("Don't launch this program as root!")
		os.Exit(1)
	}
}

func editPkgbuild() {
	if question("Do you want to edit PKGBUILD?", true) {
		edit("PKGBUILD")
	}
}

func getPackage(app string) {
	url := fmt.Sprintf(urlBase, app)
	cmd := exec.Command("git", "clone", url)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	if err := cmd.Run(); err != nil {
		os.Exit(1)
	}
}

func launchRequest(search string, withHeader bool) []byte {
	client := &http.Client{}
	req, err := http.NewRequest("GET", search, nil)
	if err != nil {
		return []byte{}
	}
	if withHeader {
		req.Header.Add(searchHeadType, searchHead)
	}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Println(err)
	}
	b, err := ioutil.ReadAll(resp.Body)
	defer resp.Body.Close()
	return b
}

func getVersion(result []byte) string {
	var pkgver, pkgrel string
	l := strings.Split(string(result), "\n")
	for _, e := range l {
		e := strings.TrimSpace(e)
		if len(e) <= 7 {
			continue
		}
		if e[:7] == "pkgver=" {
			pkgver = e[7:]
		} else if e[:7] == "pkgrel=" {
			pkgrel = e[7:]
		}
	}
	if pkgver != "" && pkgrel != "" {
		return pkgver + "-" + pkgrel
	}
	return "<unknown>"
}

func checkInstalled(app string) string {
	var localVersion string
	cmd := exec.Command("pacman", "-Q", app)
	if out, err := cmd.Output(); err == nil {
		o := strings.Fields(string(out))
		if len(o) >= 2 {
			localVersion = o[1]
		}
	}
	return localVersion
}

func searchPackage(app string, fast bool) {
	search := fmt.Sprintf(searchBase, app)
	var f interface{}
	if err := json.Unmarshal(launchRequest(search, true), &f); err != nil {
		return
	}
	j := f.(map[string]interface{})["items"]
	result := j.([]interface{})
	for _, a := range result {
		e := a.(map[string]interface{})
		n, d, s := e["name"], e["description"], e["stargazers_count"]
		i := checkInstalled(fmt.Sprintf("%v", n))
		if fast {
			if i != "" {
				fmt.Printf("\033[1m%v\033[m \033[1;36m[installed: %v]\033[m \033[1;34m(%v)\033[m\n", n, i, s)
			} else {
				fmt.Printf("\033[1m%v\033[m\033 \033[1;34m(%v)\033[m\n", n, s)
			}
		} else {
			v := string(getVersion(launchRequest(fmt.Sprintf(urlPkgbuild, n), false)))
			if i != "" {
				if v == i {
					i = " [installed]"
				} else {
					i = fmt.Sprintf(" [installed: %v]", i)
				}
			}
			fmt.Printf("\033[1m%v\033[m \033[1;32m%v\033[m\033[1;36m%v\033[m \033[1;34m(%v)\033[m\n", n, v, i, s)
		}
		fmt.Println("\t", d)
	}
}

func installPackage(app string, asdeps bool) {
	os.Chdir(tmpDir)
	wDir := tmpDir + string(os.PathSeparator) + app
	getPackage(app)
	if err := os.Chdir(wDir); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	editPkgbuild()
	var cmd *exec.Cmd
	if asdeps {
		cmd = exec.Command("makepkg", "-si", "--asdeps")
	} else {
		cmd = exec.Command("makepkg", "-si")
	}
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Run()
	os.RemoveAll(wDir)
}

func main() {
	checkUser()
	goopt.Parse(nil)
	switch {
	case *fVersion:
		fmt.Println(goopt.Version)
	case *fHelp:
		fmt.Println(goopt.Usage())
	case *fGet != "":
		getPackage(*fGet)
	case *fSearch != "":
		searchPackage(*fSearch, *fFast)
	case *fInstall != "":
		installPackage(*fInstall, *fDeps)
	default:
		fmt.Println(goopt.Usage())
	}
}
